File: compat/src/test/scala-jvm/test/scala/jdk/javaapi/CollectionConvertersTest.java
/*
 * Scala (https://www.scala-lang.org)
 *
 * Copyright EPFL and Lightbend, Inc. dba Akka
 *
 * Licensed under Apache License 2.0
 * (http://www.apache.org/licenses/LICENSE-2.0).
 *
 * See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.
 */

package test.scala.jdk.javaapi;

import org.junit.Assert;
import org.junit.Test;
import scala.Array;
import scala.Tuple2;
import scala.collection.Iterable;
import scala.collection.Iterator;
import scala.collection.concurrent.TrieMap;
import scala.collection.mutable.*;
import scala.collection.mutable.Map;
import scala.collection.mutable.Set;
import scala.jdk.javaapi.CollectionConverters;

import java.util.*;
import java.util.concurrent.ConcurrentMap;

public class CollectionConvertersTest {

    /**
     * The following conversions are supported via asScala and asJava:
     *
     * scala.collection.Iterable       <=> java.lang.Iterable
     * scala.collection.Iterator       <=> java.util.Iterator
     * scala.collection.mutable.Buffer <=> java.util.List
     * scala.collection.mutable.Set    <=> java.util.Set
     * scala.collection.mutable.Map    <=> java.util.Map
     * scala.collection.concurrent.Map <=> java.util.concurrent.ConcurrentMap
     */
    @Test
    public void shouldConvertAsScala() {
        // scala.collection.Iterable       <=> java.lang.Iterable
        java.lang.Iterable iterable = CollectionConverters.asJava(TestObjects.iterable());
        Assert.assertEquals("A", iterable.iterator().next());
        Iterable scalaIterable = CollectionConverters.asScala(iterable);
        Assert.assertEquals(TestObjects.iterable().head(), scalaIterable.head());

        // scala.collection.Iterator       <=> java.util.Iterator
        java.util.Iterator iterator = CollectionConverters.asJava(TestObjects.iterator());
        Assert.assertEquals("A", iterator.next());
        Iterator scalaIterator = CollectionConverters.asScala(iterator);
        Assert.assertTrue(scalaIterator.contains("B"));

        // scala.collection.mutable.Buffer <=> java.util.List
        List<String> list = CollectionConverters.asJava(TestObjects.buffer());
        Assert.assertEquals("A", list.get(0));
        Buffer<String> scalaBuffer = CollectionConverters.asScala(list);
        Assert.assertEquals("A", scalaBuffer.head());

        // scala.collection.mutable.Set    <=> java.util.Set
        java.util.Set<String> set = CollectionConverters.asJava(TestObjects.mutableSet());
        Assert.assertTrue(set.contains("A"));
        Set<String> scalaSet = CollectionConverters.asScala(set);
        Assert.assertTrue(scalaSet.contains("A"));

        // scala.collection.mutable.Map    <=> java.util.Map
        java.util.Map<String, String> map = CollectionConverters.asJava(TestObjects.mutableMap());
        Assert.assertEquals("B", map.get("A"));
        Map<String, String> scalaMap = CollectionConverters.asScala(map);
        Assert.assertEquals("B", scalaMap.get("A").get());

        // scala.collection.concurrent.Map <=> java.util.concurrent.ConcurrentMap
        ConcurrentMap<String, String> concurrentMap = CollectionConverters.asJava(TestObjects.concurrentMap());
        Assert.assertEquals("B", concurrentMap.get("A"));
        scala.collection.concurrent.Map<String, String> scalaConcurrentMap = CollectionConverters.asScala(concurrentMap);
        Assert.assertEquals("B", scalaConcurrentMap.get("A").get());
    }

    /**
     * The following conversions are supported via asScala and through specially-named methods to convert to Java collections, as shown:
     *
     * scala.collection.Iterable    <=> java.util.Collection   (via asJavaCollection)
     * scala.collection.Iterator    <=> java.util.Enumeration  (via asJavaEnumeration)
     * scala.collection.mutable.Map <=> java.util.Dictionary   (via asJavaDictionary)
     */
    public void convertAsCollection() {
        // scala.collection.Iterable    <=> java.util.Collection   (via asJavaCollection)
        Collection<String> collection = CollectionConverters.asJavaCollection(TestObjects.iterable());
        Assert.assertTrue(collection.contains("A"));
        Iterable<String> iterable = CollectionConverters.asScala(collection);
        Assert.assertEquals("A", iterable.head());

        // scala.collection.Iterator    <=> java.util.Enumeration  (via asJavaEnumeration)
        Enumeration<String> enumeration = CollectionConverters.asJavaEnumeration(TestObjects.iterator());
        Assert.assertEquals("A", enumeration.nextElement());
        Iterator<String> iterator = CollectionConverters.asScala(enumeration);
        Assert.assertEquals("A", iterator.next());

        // scala.collection.mutable.Map <=> java.util.Dictionary   (via asJavaDictionary)
        Dictionary<String, String> dictionary = CollectionConverters.asJavaDictionary(TestObjects.mutableMap());
        Assert.assertEquals("B", dictionary.get("A"));
        Map<String, String> map = CollectionConverters.asScala(dictionary);
        Assert.assertEquals("B", map.get("A").get());
    }

    /** In addition, the following one-way conversions are provided via asJava:
     *
     * scala.collection.Seq         => java.util.List
     * scala.collection.mutable.Seq => java.util.List
     * scala.collection.Set         => java.util.Set
     * scala.collection.Map         => java.util.Map
     */
    public void convertsAsJava() {
        // scala.collection.Seq         => java.util.List
        Assert.assertEquals("A", CollectionConverters.asJava(TestObjects.seq()).get(0));

        // scala.collection.mutable.Seq => java.util.List
        Assert.assertEquals("A", CollectionConverters.asJava(TestObjects.mutableSeq()).get(0));

        // scala.collection.Set         => java.util.Set
        Assert.assertTrue(CollectionConverters.asJava(TestObjects.set()).contains("A"));

        // scala.collection.Map         => java.util.Map
        Assert.assertEquals("B", CollectionConverters.asJava(TestObjects.map()).get("A"));
    }

    /**
     * The following one way conversion is provided via asScala:
     *
     * java.util.Properties => scala.collection.mutable.Map
     */
    public void convertsFromProperties() {
        Properties properties = new Properties();
        properties.put("key", "value");
        Map<String, String> stringStringMap = CollectionConverters.asScala(properties);
        Assert.assertEquals("value", stringStringMap.get("key").get());
    }
}

File: compat/src/test/scala/scala/jdk/CollectionConvertersTest.scala
/*
 * Scala (https://www.scala-lang.org)
 *
 * Copyright EPFL and Lightbend, Inc. dba Akka
 *
 * Licensed under Apache License 2.0
 * (http://www.apache.org/licenses/LICENSE-2.0).
 *
 * See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.
 */

package scala.jdk

import java.{lang => jl, util => ju}

import org.junit.Test

import scala.collection.compat._
import scala.collection.{mutable => m}
import scala.jdk.CollectionConverters._
import scala.{collection => c}

class CollectionConvertersTest {
  @Test
  def extensions(): Unit = {
    val it = "a b c".split(" ").iterator

    {
      val j = it.asJava
      val je = it.asJavaEnumeration
      val s = (j: ju.Iterator[String]).asScala
      assert((s: Iterator[String]) eq it)
      val es = (je: ju.Enumeration[String]).asScala
      assert((es: Iterator[String]) eq it)
    }

    {
      val i: c.Iterable[String] = it.to(Iterable)
      val j = i.asJava
      val jc = i.asJavaCollection
      val s = (j: jl.Iterable[String]).asScala
      assert((s: c.Iterable[String]) eq i)
      val cs = (jc: ju.Collection[String]).asScala
      assert((cs: c.Iterable[String]) eq i)
    }

    {
      val b: m.Buffer[String] = it.to(m.Buffer)
      val j = b.asJava
      val s = (j: ju.List[String]).asScala
      assert((s: m.Buffer[String]) eq b)
    }

    {
      val q: m.Seq[String] = it.to(m.Seq)
      val j = q.asJava
      val s = (j: ju.List[String]).asScala
      assert((s: m.Buffer[String]) == q) // not eq
    }

    {
      val q: c.Seq[String] = it.to(c.Seq)
      val j = q.asJava
      val s = (j: ju.List[String]).asScala
      assert((s: m.Buffer[String]) == q) // not eq
    }

    {
      val t: m.Set[String] = it.to(m.Set)
      val j = t.asJava
      val s = (j: ju.Set[String]).asScala
      assert((s: m.Set[String]) eq t)
    }

    {
      val t: c.Set[String] = it.to(c.Set)
      val j = t.asJava
      val s = (j: ju.Set[String]).asScala
      assert((s: m.Set[String]) == t) // not eq
    }

    {
      val p: m.Map[String, String] = m.Map(it.map(a => (a, a)).toSeq: _*)
      val j = p.asJava
      val jd = p.asJavaDictionary
      val s = (j: ju.Map[String, String]).asScala
      assert((s: m.Map[String, String]) eq p)
      val ds = (jd: ju.Dictionary[String, String]).asScala
      assert((ds: m.Map[String, String]) eq p)
    }

    {
      val p: c.Map[String, String] = c.Map(it.map(a => (a, a)).toSeq: _*)
      val j = p.asJava
      val s = (j: ju.Map[String, String]).asScala
      assert((s: m.Map[String, String]) == p) // not eq
    }

//    Scala.js doesn't like the concurrent packages
//    {
//      val p: scala.collection.concurrent.Map[String, String] = scala.collection.concurrent.TrieMap(it.map(a => (a, a)).toSeq: _*)
//      val j = p.asJava
//      val s = (j: java.util.concurrent.ConcurrentMap[String, String]).asScala
//      assert((s: scala.collection.concurrent.Map[String, String]) eq p)
//    }
  }
}
